tls/extensions/s2n_client_key_share.c

```diff
@@ -306,8 +304,128 @@ static int s2n_client_key_share_send(struct s2n_connection *conn, struct s2n_stu
+        return S2N_SUCCESS;
+    }
+
+    /* Ignore curves with unexpected share sizes */
+    if (key_share->blob.size != curve->share_size) {
+        return S2N_SUCCESS;
+    }
+
+    struct s2n_blob point_blob = { 0 };
+    GUARD(s2n_ecc_evp_read_params_point(key_share, curve->share_size, &point_blob));
+
+    /* Ignore curves with points we can't parse */
+    client_ecc_params->negotiated_curve = curve;
+    if (s2n_ecc_evp_parse_params_point(&point_blob, client_ecc_params) < 0) {
+        client_ecc_params->negotiated_curve = NULL;
+        GUARD(s2n_ecc_evp_params_free(client_ecc_params));
+        return S2N_SUCCESS;
+    }
+
+    *match = true;
+    return S2N_SUCCESS;
+}
+
+static int s2n_client_key_share_recv_pq_hybrid(struct s2n_connection *conn, struct s2n_stuffer *key_share,
+        uint16_t kem_group_iana_id, bool *match) {
+    notnull_check(conn);
+    notnull_check(key_share);
+    notnull_check(match);
+
+    const struct s2n_kem_preferences *kem_pref = NULL;
+    GUARD(s2n_connection_get_kem_preferences(conn, &kem_pref));
+    notnull_check(kem_pref);
+
+    ENSURE_POSIX(s2n_is_in_fips_mode() == false, S2N_ERR_PQ_KEMS_DISALLOWED_IN_FIPS);
+    ENSURE_POSIX(s2n_kem_preferences_includes_tls13_kem_group(kem_pref, kem_group_iana_id), S2N_ERR_KEM_UNSUPPORTED_PARAMS);
+
+    const struct s2n_kem_group *kem_group = NULL;
+    struct s2n_kem_group_params *client_kem_group_params = NULL;
+    for (size_t i = 0; i < kem_pref->tls13_kem_group_count; i++) {
+        if (kem_group_iana_id == kem_pref->tls13_kem_groups[i]->iana_id) {
+            kem_group = kem_pref->tls13_kem_groups[i];
+            client_kem_group_params = &conn->secure.client_kem_group_params[i];
+            break;
+        }
+    }
+
+    notnull_check(kem_group);
+    notnull_check(client_kem_group_params);
+
+    /* Ignore KEM groups that we've already received material for */
+    if (client_kem_group_params->kem_group) {
+        return S2N_SUCCESS;
+    }
+
+    /* Ignore KEM groups with unexpected overall total share sizes */
+    if (key_share->blob.size != kem_group->client_share_size) {
+        return S2N_SUCCESS;
+    }
+
+    uint16_t ec_share_size = 0;
+    GUARD(s2n_stuffer_read_uint16(key_share, &ec_share_size));
+    /* Ignore KEM groups with unexpected ECC share sizes */
+    if (ec_share_size != kem_group->curve->share_size) {
+        return S2N_SUCCESS;
+    }
+
+    struct s2n_blob point_blob = { 0 };
+    GUARD(s2n_ecc_evp_read_params_point(key_share, kem_group->curve->share_size, &point_blob));
+
+    /* Ignore KEM groups with ECC points we can't parse */
+    client_kem_group_params->ecc_params.negotiated_curve = kem_group->curve;
+    if (s2n_ecc_evp_parse_params_point(&point_blob, &client_kem_group_params->ecc_params) < 0) {
+        client_kem_group_params->ecc_params.negotiated_curve = NULL;
+        GUARD(s2n_ecc_evp_params_free(&client_kem_group_params->ecc_params));
+        return S2N_SUCCESS;
+    }
```

Comment:
These lines are repeated from `s2n_client_key_share_recv_ecc()` can you move them to a shared function called in both places?

Reply:
I moved the point reading/parsing into its own function. The few lines about checking the EC share size aren't truly duplicated and don't really abstract well: when receiving classic ECC shares, the share size has been read from the extension previously and stored as the blob size, and so we verify it as `key_share->blob.size != curve->share_size`; when verifying the EC share size as part of the hybrid key share, we have to read partial share size from the extension because the previously read share size was the total hybrid size (ECC + PQ).